src/wifipaf/WiFiPAFEndPoint.cpp

```diff
@@ -0,0 +1,1255 @@
+
+    // Send ack for connection handshake indication when timer expires. Sequence numbers always start at 0,
+    // and the reassembler's "last received seq num" is initialized to 0 and updated when new fragments are
+    // received from the peripheral, so we don't need to explicitly mark the ack num to send here.
+    ReturnErrorOnFailure(StartSendAckTimer());
+
+    // We've sent a capabilities request write and received a compatible response, so the connect
+    // operation has completed successfully.
+    return HandleConnectComplete();
+}
+
+// Returns number of open slots in remote receive window given the input values.
+SequenceNumber_t WiFiPAFEndPoint::AdjustRemoteReceiveWindow(SequenceNumber_t lastReceivedAck, SequenceNumber_t maxRemoteWindowSize,
+                                                            SequenceNumber_t newestUnackedSentSeqNum)
+{
+    // Assumption: SequenceNumber_t is uint8_t.
+    // Assumption: Maximum possible sequence number value is UINT8_MAX.
+    // Assumption: Sequence numbers incremented past maximum value wrap to 0.
+    // Assumption: newest unacked sent sequence number never exceeds current (and by extension, new and un-wrapped)
+    //             window boundary, so it never wraps relative to last received ack, if new window boundary would not
+    //             also wrap.
+
+    // Define new window boundary (inclusive) as uint16_t, so its value can temporarily exceed UINT8_MAX.
+    uint16_t newRemoteWindowBoundary = static_cast<uint16_t>(lastReceivedAck + maxRemoteWindowSize);
+
+    if (newRemoteWindowBoundary > UINT8_MAX && newestUnackedSentSeqNum < lastReceivedAck)
+    {
+        // New window boundary WOULD wrap, and latest unacked seq num already HAS wrapped, so add offset to difference.
+        return static_cast<uint8_t>(newRemoteWindowBoundary - (newestUnackedSentSeqNum + UINT8_MAX));
+    }
+
+    // Neither values would or have wrapped, OR new boundary WOULD wrap but latest unacked seq num does not, so no
+    // offset required.
+    return static_cast<uint8_t>(newRemoteWindowBoundary - newestUnackedSentSeqNum);
+}
+
+CHIP_ERROR WiFiPAFEndPoint::GetPktSn(Encoding::LittleEndian::Reader & reader, uint8_t * pHead, SequenceNumber_t & seqNum)
+{
+    CHIP_ERROR err;
+    BitFlags<WiFiPAFTP::HeaderFlags> rx_flags;
+    uint8_t SnOffset = 0;
+    SequenceNumber_t * pSn;
+    err = reader.Read8(rx_flags.RawStorage()).StatusCode();
+    if (rx_flags.Has(WiFiPAFTP::HeaderFlags::kHankshake))
+    {
+        // Handkshake message => No ack/sn
+        return CHIP_ERROR_INTERNAL;
+    }
+    // Always has header flag
+    SnOffset += kTransferProtocolHeaderFlagsSize;
+    if (rx_flags.Has(WiFiPAFTP::HeaderFlags::kManagementOpcode)) // Has Mgmt_Op
+    {
+        SnOffset += kTransferProtocolMgmtOpSize;
+    }
+    if (rx_flags.Has(WiFiPAFTP::HeaderFlags::kFragmentAck)) // Has ack
+    {
+        SnOffset += kTransferProtocolAckSize;
+    }
+    pSn    = pHead + SnOffset;
+    seqNum = *pSn;
```

Comment:
How do you know you didn't run off the end of the buffer here?  This is completely ignoring the actual length of the data, no?

Reply:
Yes. the checking is required